Develop connect frontend to backend (Create User, Login User, Logout user)

Casper0913 opened the pull request:
### **Jeg har tilføjet forbindelse mellem frontend og backend ved følgende:** 
- Create user
- Login user
- Logout user

### **Til som skal gøre inden man kan test**
**1. Sikre sig at man køre den nyeste database**, da det eller ville give fejl
    (Man opdatere ved brug af Dbup mappen, måske skal man droppe ens tabel inden man opdatere)
**3. At man køre både: docker (databasen), restapi og frontenden**

### **Ting som man kan teste**
1. Når man laver en bruger, kan man det og bliver den også lavet i databasen
2. Når man logger ind, kan man det og bliver der lavet en token til brugeren ind i database
3. Når man logger ud, virker det og bliver alt data omkring user i store fjernet

Casper0913 committed:
Addes fetch using rtk query for login and create user

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Fixes runninng error

Casper0913 committed:
Updates to Generate Recipes Button Handler

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Fixes paths for userApiSlice

Casper0913 committed:
Small fixes for test post

Casper0913 committed:
Save user token and username inside an cookie

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Fixes split error in cookie

Casper0913 committed:
Add logout communication between frontend and backend

Casper0913 committed:
fixes cookies for creating user + fixes response

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
code clean up + added function to retrive cookie

Casper0913 committed:
Delete addCookie.js

Casper0913 committed:
Updates to userSlice + code cleanup

Casper0913 committed:
Fixes rename of toggleswitch

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Moved user endpoints in to own file

Casper0913 committed:
Fix for error when logging out

Casper0913 committed:
Updated userSlice + fixes warnings + code clean up

Casper0913 committed:
Added comments to code + small updates to userSlice

Casper0913 committed:
Encodes input for request using encodeURIComponent

done to avoid request error for symbols

Casper0913 committed:
Code clean up for store #67

sebxGit commented on P7Internet.Frontend/react-app/src/components/ProfilePicture.jsx:
```diff
@@ -20,9 +26,37 @@ const ProfilePicture = () => {
+        }
+        
+      } catch (error) {
+        toast.error("Kunne ikke loge ud", error);
```
typo: loge --> logge

Casper0913 replied:
Sådan har fikset det i ny commit

sebxGit commented on P7Internet.Frontend/react-app/src/components/RecipeTitle.jsx:
```diff
@@ -8,12 +8,13 @@ import { pageActions } from "../features/pageSlice";
 const RecipeTitle = ({ title }) => {
   const dispatch = useDispatch();
-  const loggedInUser = useSelector((state) => state.user.loggedInUser);
+  const loggedIn = useSelector((state) => state.user.loggedInd);
```
hvorfor hedder det loggedIn et sted og loggedInd det andet sted? Foretrækker et consistent navn, og helst loggedIn, da vores program er skrevet på engelsk de andre steder ihf.

Casper0913 replied:
Har fikset dette i det ny commit.

Casper0913 committed:
type fixed + consistent name for loggedIn + small code cleanup

Lildhansen commented on P7Internet.Frontend/react-app/src/services/apiSlice.js:
```diff
@@ -0,0 +1,18 @@
+    Funktion laver en api med en basepath og header fra RTK Query, 
+    Laver et basalt request, bare med ingen endpoints, som man så injecter efter 
+    for at kunne opdele i flere filer
+*/
```
translate plz

Lildhansen replied:
gælder for alle kommentarene

Lildhansen commented on P7Internet.Frontend/react-app/src/testdata/testUser.json:
```diff
@@ -0,0 +1,5 @@
+    "id": "3a831c1c-5a2a-4417-821c-37e36d763db4",
+    "name": "u",
+    "emailAddress": "user1@mail.dk"
+}
```
bliver det her brugt til tests nogen steder eller hvordan

Lildhansen commented on P7Internet.Frontend/react-app/src/helperFunctions/retriveCookie.js:
```diff
@@ -0,0 +1,13 @@
+//Function for retriving data from cookies
+const retriveCookie = (retriveValue) => {
```
synes godt denne funktion kunne testes.
måske hvis man selv skal passe cookies med som argument, og så kan man mocke dem i testen

Lildhansen commented on P7Internet.Frontend/react-app/src/components/GenerateRecipeButton.jsx:
```diff
@@ -1,10 +1,11 @@
-import "react-toastify/dist/ReactToastify.css";
+import { toast } from 'react-toastify';
+import 'react-toastify/dist/ReactToastify.css';
+//import { recipeGenerationActions } from "../features/recipeGenerationSlice";
```
plz delete

Lildhansen commented on P7Internet.Frontend/react-app/src/components/LoginBox.jsx:
```diff
@@ -1,40 +1,69 @@
+
+  //Functions is async because it needs to wait for the response from the backend
+  const handleLogin = async () => {
+    //Functions from redux that allows to test weather error or loading
```
synes kommentaren skal sigehvad if'en betyder, fx (uden jeg helt ved hvad den gør):
if login is not loading or there are no errors when logging in

Lildhansen commented on P7Internet.Frontend/react-app/src/components/LoginBox.jsx:
```diff
@@ -80,15 +107,50 @@ const LoginBox = ({ closeModal }) => {
+            setPassword("");
+          }
+        } catch (error) {
+          toast.error("Kodeordet eller brugernavnet er indtastet forkert");
```
det skal ændres til noget andet. burde det overhovedet være en toast, det kan vel ikke være brugerens skyld hvis vi ender her eller hvad?

Lildhansen commented on P7Internet.Frontend/react-app/src/components/LoginBox.jsx:
```diff
@@ -80,15 +107,50 @@ const LoginBox = ({ closeModal }) => {
-      setPassword("");
-      toast.success("Indsæt nu dine oplysninger for at logge ind.");
-      toast.success("Din bruger er nu tilføjet til databasen!");
+      //Functions from redux that allows to test weather error or loading
```
samme comment som ovenover

Lildhansen commented on P7Internet.Frontend/react-app/src/components/LoginBox.jsx:
```diff
@@ -80,15 +107,50 @@ const LoginBox = ({ closeModal }) => {
         "Kodeordet skal bestå af mindst et tal, et stort bogstav, et lille bogstav og være mellem 6 og 20 tegn langt uden brug af specielle tegn."
       );
     else {
```
der er ret meget kodeduplikering. kunne det komme ud i funktioner eventuelt

Lildhansen replied:
altså login og create user

Lildhansen commented on P7Internet.Frontend/react-app/src/components/ProfilePicture.jsx:
```diff
@@ -20,9 +26,37 @@ const ProfilePicture = () => {
+      const sessionToken = retriveCookie('sessionToken='); 
+      const userId = retriveCookie('userid=');
+
+      //console.log(`Retrived tokens: UserId: ${userId}, sessionToken: ${sessionToken}`);
```
remove plz

Lildhansen commented on P7Internet.Frontend/react-app/src/components/ProfilePicture.jsx:
```diff
@@ -20,9 +26,37 @@ const ProfilePicture = () => {
+        // Waits for the response and allows to use response (unwrap, because JSON)
+        const response = await userLogOut({ userId: encodedUserId, sessionToken: encodedSessionToken });
+        if(response){
+          //console.log(response);
```
remove plz

Lildhansen commented on P7Internet.Frontend/react-app/src/components/ProfilePicture.jsx:
```diff
@@ -20,9 +26,37 @@ const ProfilePicture = () => {
+        }
+        
+      } catch (error) {
+        toast.error("Kunne ikke logge ud", error);
```
printer vi erroren nu til brugeren. det lyder ikke særlig repræsentabelt for brugeren

Lildhansen replied:
det samme for toasten med ikke kunne lave en bruger. synes ikke det burde være en toast men at vi regel håndterer det fordi det har vel intet med brugeren at gøre

Lildhansen commented on P7Internet.Frontend/react-app/src/components/ProfilePicture.jsx:
```diff
@@ -20,9 +26,37 @@ const ProfilePicture = () => {
+        const response = await userLogOut({ userId: encodedUserId, sessionToken: encodedSessionToken });
+        if(response){
+          //console.log(response);
+          toast.success("Logget ud");
```
kunne det ikke stå mere som en sætning.
fx "du loggede succesfuldt ud" eller såen

Lildhansen commented on P7Internet.Frontend/react-app/src/components/ProfilePicture.jsx:
```diff
@@ -20,9 +26,37 @@ const ProfilePicture = () => {
     setOpen(false);
```
what is this

Lildhansen replied:
du har også tilføjet true senere

Lildhansen commented on P7Internet.Frontend/react-app/src/components/RecipeTitle.jsx:
```diff
@@ -8,12 +8,13 @@ import { pageActions } from "../features/pageSlice";
     event.preventDefault();
-    if (loggedInUser.length !== 1) {
+    if (loggedIn === false) {
       //if not logged in
```
nu når du har ændret if'en synes jeg ikke comment er nødvendig

Lildhansen commented on P7Internet.Frontend/react-app/src/components/RecipeTitle.jsx:
```diff
@@ -30,8 +31,9 @@ const RecipeTitle = ({ title }) => {
-      if (loggedInUser.length !== 1) return;
-      if (loggedInUser[0]["heartedRecipes"].includes(title)) {
+      if (heartedRecipes === undefined){
+        //console.log(heartedRecipes);
```
yeet plz

Lildhansen replied:
hele if'en vel, fordi vi har ingen error handling for det (ellers skal der tilføjes error handling)

Lildhansen committed:
remove all button fix

Lildhansen commented on P7Internet.Frontend/react-app/src/features/userSlice.js:
```diff
@@ -1,53 +1,31 @@
     loginUser(state, action) {
-      state.loggedInUser = action.payload;
+      state.user = action.payload;
+      state.heartedRecipes = [];
```
det er lidt wack vi resetter hearted recipes når man logger ind.

Lildhansen commented on P7Internet.Frontend/react-app/src/helperFunctions/generateOpenAiString.js:
```diff
@@ -0,0 +1,27 @@
+
+//Skal fjernes når endpoint er klar. Blev lavet for at kunne test OpenAi
```
plz remember to remove

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Merge branch 'develop-connect_frontend_to_backend' of https://github.com/Jaffenheimer/p7-internet into develop-connect_frontend_to_backend

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Fixes comments to pr

Casper0913 committed:
Added test for cookie

Casper0913 committed:
Formating + Changes comments to english

Casper0913 committed:
Ændre tekst i toast

Casper0913 committed:
Update LoginBox.jsx

Casper0913 committed:
Fixes serializable warning when testing

Lildhansen committed:
more cleanup

Casper0913 committed:
Moved input validation functions into a file

Casper0913 committed:
Combined Create and Login user + add cookie functions + codeclean  up  + more

- Split Create user and login user into one function
- Added function to add and delete cookies
- Moved getLocation into file
- Moved inputValidation into file
- Code Cleanup

Casper0913 committed:
Update ProfilePicture.jsx

Casper0913 committed:
Merge branch 'develop' into develop-connect_frontend_to_backend

Casper0913 committed:
Added test for cookiehandler

Casper0913 committed:
Removed axios from project + addeding loading state to login/create user function

- Also removed example file from project

Casper0913 committed:
prettier files worked on + code cleanup

Casper0913 committed:
added printing error when creating and login in
